<a href="https://colab.research.google.com/github/archisha13/Deep-Learning/blob/main/B072_DL_lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



>              NAME: ARCHISHA SRIVASTAVA 



>              DL_LAB3

>              ROLL NUM: B072


AIM: Using backpropagation to reduce error

In [ ]:
import numpy as np
import pandas as pd
import tensorflow as tf
# tensorflow is used to build the network
import matplotlib as plt

In [ ]:
#reading the csv file
df = pd.read_csv("/content/iris_data (1).csv")
df.head()

sepal_length  sepal_width  petal_length  petal_width      species
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

In [ ]:
df.shape

(150, 5)

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [ ]:
df['species'].unique()
# we can see there are  3 types of flowers 

array(['Iris-setosa', 'Iris-versicolor', 'Iris-virginica'], dtype=object)

In [ ]:
#converting categorical variables to numerical value
#df1 =pd.get_dummies(df["species"])
from sklearn.preprocessing import OneHotEncoder
#One hot Encoder and get dummies does the same work

In [ ]:
#converting the dataframe to numpy array
#extracting all the rows and (-1) last column
label = np.array(df.iloc[:,-1])
print(label)
label.shape
#(150,) means it is a vector[row vector]
#if  (,150) column vector


['Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa' 'Iris-setosa'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor'
 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor' 'Iris-versicolor

(150,)

In [ ]:
#number of rows taken automatically so take -1
#(-1,1)will change the vectors of any size to matrix
label=label.reshape((-1,1))
print(label)
label.shape

[['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-setosa']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['Iris-versicolor']
 ['

(150, 1)

In [ ]:
#sparse =False returns an array otherwise will return sparse matrix
one_hot = OneHotEncoder(sparse=False)
y_onehot =one_hot.fit_transform(label)
y_onehot.shape

(150, 3)

In [ ]:
#for dataframe we use head and this is numpy array
y_onehot

array([[1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0., 0.],
       [1., 0

In [ ]:
dataset = pd.DataFrame({'folwer1':y_onehot[:,0],'folwer2':y_onehot[:,1],'folwer3':y_onehot[:,2]})

In [ ]:
dataset.head()

folwer1  folwer2  folwer3
0      1.0      0.0      0.0
1      1.0      0.0      0.0
2      1.0      0.0      0.0
3      1.0      0.0      0.0
4      1.0      0.0      0.0

In [ ]:
dataset.tail()

folwer1  folwer2  folwer3
145      0.0      0.0      1.0
146      0.0      0.0      1.0
147      0.0      0.0      1.0
148      0.0      0.0      1.0
149      0.0      0.0      1.0

In [ ]:
#dropping species column and adding the one_hot_enocoded values
#axis=1 drop the column with given name
df.drop('species',axis=1,inplace=True)

In [ ]:
df.head()

sepal_length  sepal_width  petal_length  petal_width
0           5.1          3.5           1.4          0.2
1           4.9          3.0           1.4          0.2
2           4.7          3.2           1.3          0.2
3           4.6          3.1           1.5          0.2
4           5.0          3.6           1.4          0.2

In [ ]:
df_final= pd.concat([df,dataset],axis=1)
df_final.head()

sepal_length  sepal_width  petal_length  petal_width  folwer1  folwer2  \
0           5.1          3.5           1.4          0.2      1.0      0.0   
1           4.9          3.0           1.4          0.2      1.0      0.0   
2           4.7          3.2           1.3          0.2      1.0      0.0   
3           4.6          3.1           1.5          0.2      1.0      0.0   
4           5.0          3.6           1.4          0.2      1.0      0.0   

   folwer3  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [ ]:
std_scalar=StandardScaler()
df_final.iloc[ : ,[0,1,2,3]]=std_scalar.fit_transform(df_final.iloc[:,[0,1,2,3]])

In [ ]:
df_final.head()

sepal_length  sepal_width  petal_length  petal_width  folwer1  folwer2  \
0     -0.900681     1.032057     -1.341272    -1.312977      1.0      0.0   
1     -1.143017    -0.124958     -1.341272    -1.312977      1.0      0.0   
2     -1.385353     0.337848     -1.398138    -1.312977      1.0      0.0   
3     -1.506521     0.106445     -1.284407    -1.312977      1.0      0.0   
4     -1.021849     1.263460     -1.341272    -1.312977      1.0      0.0   

   folwer3  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      0.0

In [ ]:
#sigmoid function
def sig(x):
  sig_x= 1/(1+np.exp(-x))
  return sig_x

In [ ]:
def accuracy(y_pred,y_true):
  #if both pred and actual values are same those will be counted
  acc = y_pred.argmax(axis=1)==y_true.argmax(axis=1)
  #in the row which column is taking max value i.e 1 take that row so from predicted value we check if 1 is present in the same col
  temp = acc.mean()
  return temp

In [ ]:
#mean square error
def m_s_e(y_pred,y_true):
  temp=(((y_pred-y_true)**2).sum())/y_pred.size
  return temp

In [ ]:
input = df_final.iloc[:,0:4]

In [ ]:
N=input.size
iterations=100
learning_rate = 0.01
input_size=4
hidden_size =2
output_size = 3
result1 =pd.DataFrame(columns=['mse','accuracy'])

In [ ]:
np.random.seed(10)
#np.random.seed(40)
'''it doesnt matter if we chNG'''
w1 = np.random.normal(scale =0.5,size =(input_size,hidden_size))
w2 = np.random.normal(scale =0.5,size =(hidden_size,output_size))
#for first neuron of hidden layer
bias1 = np.random.rand(1)
#for second neuron of hidden layer
bias2 = np.random.rand(1)

backpropagation

In [ ]:
'''reason why bias wasnt included????????????????'''
for i in range(iterations):
  #output of hidden layer
  z1 =  np.dot(input,w1)+bias1
  #output at hidden layer(1) in sigmoid fun is a1
  a1=sig(z1)
  #output at output layer(layer2)
  z2 =  np.dot(a1,w2)
  a2=sig(z2)
  #a2 is the predicted value(output)
  #y_onehot is the actual value
  #calculating error
  mse =m_s_e(a2,y_onehot)
  #accuracy
  acc= accuracy(a2,y_onehot)
  #appending these updated value in our df to check the changes in error so accordingly will change weights
  result1=result1.append({'mse':mse,'accuracy':acc},ignore_index=True)
  #error btw output layer and predicted
  E1= a2-y_onehot
  #formula for partial derivative of output
  dw1 = E1*a2*(1-a2)#pd
  E2=np.dot(dw1,w2.T)#error
  dw2=E2*a1*(1-a1)#pd
  
  w2_update = np.dot(a1.T,dw1)/N
  w1_update =np.dot(input.T,dw2)/N
  
  w2=w2-learning_rate*w2_update
  w1=w1-learning_rate*w1_update



In [ ]:
result1.shape
#100 rows and 2 columns
#for each iterations mse and acc are the 2 cols

(100, 2)

In [ ]:
result1.max()#100 iterations for learning rate 0.01

mse         0.274659
accuracy    0.513333
dtype: float64

In [ ]:
result1.max()#2000 iterations for learning rate 0.01

mse         0.274659
accuracy    0.513333
dtype: float64

In [ ]:
result1.max()#5000 iterations for learning rate 0.01

mse         0.274659
accuracy    0.560000
dtype: float64

In [ ]:
result1.max()#9000 iterations for learning rate 0.01

mse         0.274659
accuracy    0.660000
dtype: float64

In [ ]:
result1.max()#2000 iterations for learning rate 0.1

mse         0.274659
accuracy    0.666667
dtype: float64

In [ ]:
result1.max()#5000 iterations for learning rate 0.1

mse         0.274659
accuracy    0.886667
dtype: float64

In [ ]:
result1.max()#9000 iterations for learning rate 0.1

mse         0.274659
accuracy    0.940000
dtype: float64

# CONCLUSION
A 3-layer network with 1 hidden layer is built for classification of iris dataset

*   The given network has only 2 neurons in hidden layer

*  For learning rate =0.01 
*  For 100 iterations mse is 0.274659 and acc is 0.513333
*  For 2000 iterations mse is 0.274659 and acc is 0.513333
*  For 5000 iterations mse is 0.274659 and acc is 0.560000
*  For 9000 iterations mse is  0.274659 and acc is 0.660000


>  For increased learning rate of 0.1
  * If number of iterations is increased 
  * For 2000 iterations the mse 0.274659 and acc is 0.666667 
  * For 5000 iterations mse is 0.274659 and acc is 0.886667
  * For 9000 iterations mse is 0.274659 and acc is .940000

Since the network is too simple the accuracy is very low ,to improve the accuracy ,the number of hidden layer and the number of neurons in each layer should be increased



> If random state of initialization of weights(Random seed value) is changed to another value there is no appreciable change in error and accuracy.
This is because network can be initialized to any value but it is the back propagation algorithm which changes the values to reach the optimum value




